File: 2022/day22/c/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_ROWS 256
#define MAX_COLS 256
#define MAX_PATH 10000
#define MAX_LINE 8192

typedef struct {
    int is_number;
    int value;    // number of steps or direction: 'L' or 'R'
} Instruction;

char grid[MAX_ROWS][MAX_COLS];
int height = 0;
int width = 0;
Instruction instructions[MAX_PATH];
int num_instructions = 0;

// Directions: 0=right, 1=down, 2=left, 3=up
int DR[] = {0, 1, 0, -1};
int DC[] = {1, 0, -1, 0};

void parse_input(const char *filename) {
    FILE *f = fopen(filename, "r");
    if (!f) {
        perror("Failed to open input file");
        exit(1);
    }

    char line[MAX_LINE];

    // Read grid lines
    while (fgets(line, sizeof(line), f)) {
        // Check for empty line (separator between grid and path)
        if (line[0] == '\n' || line[0] == '\r') {
            break;
        }

        int len = strlen(line);
        // Remove newline
        while (len > 0 && (line[len-1] == '\n' || line[len-1] == '\r')) {
            line[--len] = '\0';
        }

        // Copy line to grid, tracking width
        strcpy(grid[height], line);
        if (len > width) {
            width = len;
        }
        height++;
    }

    // Pad all rows to the same width
    for (int r = 0; r < height; r++) {
        int len = strlen(grid[r]);
        while (len < width) {
            grid[r][len++] = ' ';
        }
        grid[r][len] = '\0';
    }

    // Read path line
    if (fgets(line, sizeof(line), f)) {
        int len = strlen(line);
        while (len > 0 && (line[len-1] == '\n' || line[len-1] == '\r')) {
            line[--len] = '\0';
        }

        // Parse path into instructions
        int i = 0;
        while (i < len) {
            if (isdigit(line[i])) {
                int num = 0;
                while (i < len && isdigit(line[i])) {
                    num = num * 10 + (line[i] - '0');
                    i++;
                }
                instructions[num_instructions].is_number = 1;
                instructions[num_instructions].value = num;
                num_instructions++;
            } else {
                instructions[num_instructions].is_number = 0;
                instructions[num_instructions].value = line[i];
                num_instructions++;
                i++;
            }
        }
    }

    fclose(f);
}

int part1(void) {
    // Find starting position (leftmost open tile on top row)
    int row = 0, col = 0;
    while (grid[0][col] != '.') col++;
    int facing = 0;

    for (int i = 0; i < num_instructions; i++) {
        if (instructions[i].is_number) {
            int steps = instructions[i].value;
            for (int s = 0; s < steps; s++) {
                int dr = DR[facing], dc = DC[facing];
                int nr = row + dr, nc = col + dc;

                // Wrap around if needed
                if (facing == 0) { // Right
                    if (nc >= width || grid[nr][nc] == ' ') {
                        nc = 0;
                        while (grid[nr][nc] == ' ') nc++;
                    }
                } else if (facing == 2) { // Left
                    if (nc < 0 || grid[nr][nc] == ' ') {
                        nc = width - 1;
                        while (grid[nr][nc] == ' ') nc--;
                    }
                } else if (facing == 1) { // Down
                    if (nr >= height || grid[nr][nc] == ' ') {
                        nr = 0;
                        while (grid[nr][nc] == ' ') nr++;
                    }
                } else if (facing == 3) { // Up
                    if (nr < 0 || grid[nr][nc] == ' ') {
                        nr = height - 1;
                        while (grid[nr][nc] == ' ') nr--;
                    }
                }

                // Check if we hit a wall
                if (grid[nr][nc] == '#') {
                    break;
                }

                row = nr;
                col = nc;
            }
        } else {
            // Turn
            if (instructions[i].value == 'R') {
                facing = (facing + 1) % 4;
            } else {
                facing = (facing + 3) % 4;  // -1 mod 4
            }
        }
    }

    return 1000 * (row + 1) + 4 * (col + 1) + facing;
}

// Get cube face and local coordinates for the actual input layout:
//    12
//    3
//   45
//   6
void get_cube_face(int row, int col, int face_size, int *face, int *lr, int *lc) {
    int face_row = row / face_size;
    int face_col = col / face_size;
    *lr = row % face_size;
    *lc = col % face_size;

    if (face_row == 0 && face_col == 1) *face = 1;
    else if (face_row == 0 && face_col == 2) *face = 2;
    else if (face_row == 1 && face_col == 1) *face = 3;
    else if (face_row == 2 && face_col == 0) *face = 4;
    else if (face_row == 2 && face_col == 1) *face = 5;
    else if (face_row == 3 && face_col == 0) *face = 6;
    else *face = -1;
}

// Wrap on cube edges for the specific layout
void wrap_cube(int row, int col, int facing, int face_size, int *nr, int *nc, int *nf) {
    int S = face_size;
    int face, lr, lc;
    get_cube_face(row, col, S, &face, &lr, &lc);

    // Default: no change
    *nr = row;
    *nc = col;
    *nf = facing;

    if (face == 1) {
        if (facing == 3) { // Up: goes to face 6, from left, facing right
            *nr = 3*S + lc;
            *nc = 0;
            *nf = 0;
        } else if (facing == 2) { // Left: goes to face 4, from left, facing right (inverted)
            *nr = 3*S - 1 - lr;
            *nc = 0;
            *nf = 0;
        }
    } else if (face == 2) {
        if (facing == 0) { // Right: goes to face 5, from right, facing left (inverted)
            *nr = 3*S - 1 - lr;
            *nc = 2*S - 1;
            *nf = 2;
        } else if (facing == 1) { // Down: goes to face 3, from right, facing left
            *nr = S + lc;
            *nc = 2*S - 1;
            *nf = 2;
        } else if (facing == 3) { // Up: goes to face 6, from bottom, facing up
            *nr = 4*S - 1;
            *nc = lc;
            *nf = 3;
        }
    } else if (face == 3) {
        if (facing == 0) { // Right: goes to face 2, from bottom, facing up
            *nr = S - 1;
            *nc = 2*S + lr;
            *nf = 3;
        } else if (facing == 2) { // Left: goes to face 4, from top, facing down
            *nr = 2*S;
            *nc = lr;
            *nf = 1;
        }
    } else if (face == 4) {
        if (facing == 3) { // Up: goes to face 3, from left, facing right
            *nr = S + lc;
            *nc = S;
            *nf = 0;
        } else if (facing == 2) { // Left: goes to face 1, from left, facing right (inverted)
            *nr = S - 1 - lr;
            *nc = S;
            *nf = 0;
        }
    } else if (face == 5) {
        if (facing == 0) { // Right: goes to face 2, from right, facing left (inverted)
            *nr = S - 1 - lr;
            *nc = 3*S - 1;
            *nf = 2;
        } else if (facing == 1) { // Down: goes to face 6, from right, facing left
            *nr = 3*S + lc;
            *nc = S - 1;
            *nf = 2;
        }
    } else if (face == 6) {
        if (facing == 0) { // Right: goes to face 5, from bottom, facing up
            *nr = 3*S - 1;
            *nc = S + lr;
            *nf = 3;
        } else if (facing == 1) { // Down: goes to face 2, from top, facing down
            *nr = 0;
            *nc = 2*S + lc;
            *nf = 1;
        } else if (facing == 2) { // Left: goes to face 1, from top, facing down
            *nr = 0;
            *nc = S + lr;
            *nf = 1;
        }
    }
}

int part2(void) {
    // Determine face size
    int face_size = (height > 50) ? 50 : 4;

    // Find starting position
    int row = 0, col = 0;
    while (grid[0][col] != '.') col++;
    int facing = 0;

    for (int i = 0; i < num_instructions; i++) {
        if (instructions[i].is_number) {
            int steps = instructions[i].value;
            for (int s = 0; s < steps; s++) {
                int dr = DR[facing], dc = DC[facing];
                int nr = row + dr, nc = col + dc;
                int nf = facing;

                // Check if we need to wrap
                int need_wrap = 0;
                if (nr < 0 || nr >= height || nc < 0 || nc >= width) {
                    need_wrap = 1;
                } else if (grid[nr][nc] == ' ') {
                    need_wrap = 1;
                }

                if (need_wrap) {
                    wrap_cube(row, col, facing, face_size, &nr, &nc, &nf);
                }

                // Check if we hit a wall
                if (grid[nr][nc] == '#') {
                    break;
                }

                row = nr;
                col = nc;
                facing = nf;
            }
        } else {
            // Turn
            if (instructions[i].value == 'R') {
                facing = (facing + 1) % 4;
            } else {
                facing = (facing + 3) % 4;
            }
        }
    }

    return 1000 * (row + 1) + 4 * (col + 1) + facing;
}

int main(void) {
    parse_input("../input.txt");

    printf("Part 1: %d\n", part1());
    printf("Part 2: %d\n", part2());

    return 0;
}

File: 2022/day22/zig/solution.zig
const std = @import("std");

const Direction = enum(u8) {
    right = 0,
    down = 1,
    left = 2,
    up = 3,

    fn turn_right(self: Direction) Direction {
        return @enumFromInt((@intFromEnum(self) + 1) % 4);
    }

    fn turn_left(self: Direction) Direction {
        return @enumFromInt((@intFromEnum(self) + 3) % 4);
    }

    fn dr(self: Direction) i32 {
        const deltas = [_]i32{ 0, 1, 0, -1 };
        return deltas[@intFromEnum(self)];
    }

    fn dc(self: Direction) i32 {
        const deltas = [_]i32{ 1, 0, -1, 0 };
        return deltas[@intFromEnum(self)];
    }
};

const Instruction = union(enum) {
    move: u32,
    turn_left: void,
    turn_right: void,
};

const MAX_INSTRUCTIONS = 10000;
const MAX_ROWS = 300;

fn parse_instructions(path: []const u8, out: *[MAX_INSTRUCTIONS]Instruction) usize {
    var count: usize = 0;
    var i: usize = 0;

    while (i < path.len and count < MAX_INSTRUCTIONS) {
        if (path[i] >= '0' and path[i] <= '9') {
            var j = i;
            while (j < path.len and path[j] >= '0' and path[j] <= '9') {
                j += 1;
            }
            const num = std.fmt.parseInt(u32, path[i..j], 10) catch 0;
            out[count] = .{ .move = num };
            count += 1;
            i = j;
        } else if (path[i] == 'L') {
            out[count] = .turn_left;
            count += 1;
            i += 1;
        } else if (path[i] == 'R') {
            out[count] = .turn_right;
            count += 1;
            i += 1;
        } else {
            i += 1;
        }
    }

    return count;
}

const Grid = struct {
    rows: [MAX_ROWS][]const u8,
    row_count: usize,
    width: usize,

    fn get_cell(self: *const Grid, row: i32, col: i32) u8 {
        if (row < 0 or col < 0) return ' ';
        const r: usize = @intCast(row);
        const c: usize = @intCast(col);
        if (r >= self.row_count) return ' ';
        if (c >= self.rows[r].len) return ' ';
        return self.rows[r][c];
    }
};

fn part1(grid: *const Grid, instructions: []const Instruction) i64 {
    const height: i32 = @intCast(grid.row_count);
    const width: i32 = @intCast(grid.width);

    // Find starting position
    var col: i32 = 0;
    for (grid.rows[0], 0..) |c, idx| {
        if (c == '.') {
            col = @intCast(idx);
            break;
        }
    }
    var row: i32 = 0;
    var facing = Direction.right;

    for (instructions) |instr| {
        switch (instr) {
            .move => |steps| {
                for (0..steps) |_| {
                    const dr = facing.dr();
                    const dc = facing.dc();
                    var nr = row + dr;
                    var nc = col + dc;

                    // Wrap around if needed
                    if (facing == .right) {
                        if (nc >= width or grid.get_cell(nr, nc) == ' ') {
                            nc = 0;
                            while (grid.get_cell(nr, nc) == ' ') {
                                nc += 1;
                            }
                        }
                    } else if (facing == .left) {
                        if (nc < 0 or grid.get_cell(nr, nc) == ' ') {
                            nc = width - 1;
                            while (grid.get_cell(nr, nc) == ' ') {
                                nc -= 1;
                            }
                        }
                    } else if (facing == .down) {
                        if (nr >= height or grid.get_cell(nr, nc) == ' ') {
                            nr = 0;
                            while (grid.get_cell(nr, nc) == ' ') {
                                nr += 1;
                            }
                        }
                    } else if (facing == .up) {
                        if (nr < 0 or grid.get_cell(nr, nc) == ' ') {
                            nr = height - 1;
                            while (grid.get_cell(nr, nc) == ' ') {
                                nr -= 1;
                            }
                        }
                    }

                    // Check if we hit a wall
                    if (grid.get_cell(nr, nc) == '#') {
                        break;
                    }

                    row = nr;
                    col = nc;
                }
            },
            .turn_left => {
                facing = facing.turn_left();
            },
            .turn_right => {
                facing = facing.turn_right();
            },
        }
    }

    return 1000 * (row + 1) + 4 * (col + 1) + @as(i64, @intFromEnum(facing));
}

fn get_cube_face(row: i32, col: i32, face_size: i32) struct { face: i32, lr: i32, lc: i32 } {
    const face_row = @divTrunc(row, face_size);
    const face_col = @divTrunc(col, face_size);
    const lr = @mod(row, face_size);
    const lc = @mod(col, face_size);

    // Cube layout:
    //   12
    //   3
    //  45
    //  6
    const face: i32 = if (face_row == 0 and face_col == 1) 1
    else if (face_row == 0 and face_col == 2) 2
    else if (face_row == 1 and face_col == 1) 3
    else if (face_row == 2 and face_col == 0) 4
    else if (face_row == 2 and face_col == 1) 5
    else if (face_row == 3 and face_col == 0) 6
    else -1;

    return .{ .face = face, .lr = lr, .lc = lc };
}

fn wrap_cube(row: i32, col: i32, facing: Direction, face_size: i32) struct { nr: i32, nc: i32, nf: Direction } {
    const S = face_size;
    const cube_info = get_cube_face(row, col, S);
    const face = cube_info.face;
    const lr = cube_info.lr;
    const lc = cube_info.lc;

    // Wrapping rules for the actual input layout
    if (face == 1) {
        if (facing == .up) {
            // Goes to face 6, from left, facing right
            return .{ .nr = 3 * S + lc, .nc = 0, .nf = .right };
        } else if (facing == .left) {
            // Goes to face 4, from left, facing right (inverted)
            return .{ .nr = 3 * S - 1 - lr, .nc = 0, .nf = .right };
        }
    } else if (face == 2) {
        if (facing == .right) {
            // Goes to face 5, from right, facing left (inverted)
            return .{ .nr = 3 * S - 1 - lr, .nc = 2 * S - 1, .nf = .left };
        } else if (facing == .down) {
            // Goes to face 3, from right, facing left
            return .{ .nr = S + lc, .nc = 2 * S - 1, .nf = .left };
        } else if (facing == .up) {
            // Goes to face 6, from bottom, facing up
            return .{ .nr = 4 * S - 1, .nc = lc, .nf = .up };
        }
    } else if (face == 3) {
        if (facing == .right) {
            // Goes to face 2, from bottom, facing up
            return .{ .nr = S - 1, .nc = 2 * S + lr, .nf = .up };
        } else if (facing == .left) {
            // Goes to face 4, from top, facing down
            return .{ .nr = 2 * S, .nc = lr, .nf = .down };
        }
    } else if (face == 4) {
        if (facing == .up) {
            // Goes to face 3, from left, facing right
            return .{ .nr = S + lc, .nc = S, .nf = .right };
        } else if (facing == .left) {
            // Goes to face 1, from left, facing right (inverted)
            return .{ .nr = S - 1 - lr, .nc = S, .nf = .right };
        }
    } else if (face == 5) {
        if (facing == .right) {
            // Goes to face 2, from right, facing left (inverted)
            return .{ .nr = S - 1 - lr, .nc = 3 * S - 1, .nf = .left };
        } else if (facing == .down) {
            // Goes to face 6, from right, facing left
            return .{ .nr = 3 * S + lc, .nc = S - 1, .nf = .left };
        }
    } else if (face == 6) {
        if (facing == .right) {
            // Goes to face 5, from bottom, facing up
            return .{ .nr = 3 * S - 1, .nc = S + lr, .nf = .up };
        } else if (facing == .down) {
            // Goes to face 2, from top, facing down
            return .{ .nr = 0, .nc = 2 * S + lc, .nf = .down };
        } else if (facing == .left) {
            // Goes to face 1, from top, facing down
            return .{ .nr = 0, .nc = S + lr, .nf = .down };
        }
    }

    // Shouldn't reach here
    return .{ .nr = row, .nc = col, .nf = facing };
}

fn part2(grid: *const Grid, instructions: []const Instruction) i64 {
    const height: i32 = @intCast(grid.row_count);
    const width: i32 = @intCast(grid.width);

    // Determine face size
    const face_size: i32 = if (height > 50) 50 else 4;

    // Find starting position
    var col: i32 = 0;
    for (grid.rows[0], 0..) |c, idx| {
        if (c == '.') {
            col = @intCast(idx);
            break;
        }
    }
    var row: i32 = 0;
    var facing = Direction.right;

    for (instructions) |instr| {
        switch (instr) {
            .move => |steps| {
                for (0..steps) |_| {
                    const dr = facing.dr();
                    const dc = facing.dc();
                    var nr = row + dr;
                    var nc = col + dc;
                    var nf = facing;

                    // Check if we need to wrap
                    var need_wrap = false;
                    if (nr < 0 or nr >= height or nc < 0 or nc >= width) {
                        need_wrap = true;
                    } else if (grid.get_cell(nr, nc) == ' ') {
                        need_wrap = true;
                    }

                    if (need_wrap) {
                        const wrap_result = wrap_cube(row, col, facing, face_size);
                        nr = wrap_result.nr;
                        nc = wrap_result.nc;
                        nf = wrap_result.nf;
                    }

                    // Check if we hit a wall
                    if (grid.get_cell(nr, nc) == '#') {
                        break;
                    }

                    row = nr;
                    col = nc;
                    facing = nf;
                }
            },
            .turn_left => {
                facing = facing.turn_left();
            },
            .turn_right => {
                facing = facing.turn_right();
            },
        }
    }

    return 1000 * (row + 1) + 4 * (col + 1) + @as(i64, @intFromEnum(facing));
}

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    // Read input file
    const file = std.fs.cwd().openFile("../input.txt", .{}) catch |err| {
        std.debug.print("Failed to open ../input.txt: {}\n", .{err});
        return err;
    };
    defer file.close();

    const content = try file.readToEndAlloc(allocator, 1024 * 1024);
    defer allocator.free(content);

    // Split into grid and path
    var parts = std.mem.splitSequence(u8, content, "\n\n");
    const grid_text = parts.first();
    const path_text = std.mem.trim(u8, parts.rest(), "\n\r ");

    // Parse grid
    var grid: Grid = undefined;
    grid.row_count = 0;
    grid.width = 0;

    var lines = std.mem.splitScalar(u8, grid_text, '\n');
    while (lines.next()) |line| {
        if (grid.row_count < MAX_ROWS) {
            grid.rows[grid.row_count] = line;
            if (line.len > grid.width) grid.width = line.len;
            grid.row_count += 1;
        }
    }

    // Parse instructions
    var instructions: [MAX_INSTRUCTIONS]Instruction = undefined;
    const instr_count = parse_instructions(path_text, &instructions);

    const p1 = part1(&grid, instructions[0..instr_count]);
    const p2 = part2(&grid, instructions[0..instr_count]);

    // Use posix write to stdout
    const stdout_fd = std.posix.STDOUT_FILENO;
    var buf1: [64]u8 = undefined;
    const msg1 = std.fmt.bufPrint(&buf1, "Part 1: {d}\n", .{p1}) catch unreachable;
    _ = std.posix.write(stdout_fd, msg1) catch {};

    var buf2: [64]u8 = undefined;
    const msg2 = std.fmt.bufPrint(&buf2, "Part 2: {d}\n", .{p2}) catch unreachable;
    _ = std.posix.write(stdout_fd, msg2) catch {};
}
